### PR TITLE
replace ml with hosted_ml

### PR DESCRIPTION
**Pull request by haydnv**

### CODE REVIEW

**Review by KOHbDS — COMMENTED** [review]
I don't think `Variable` should have a `gradients` method, because `Variable.gradients` exists only to calculate a gradient of the same other shape using `Transform`. Maybe it is better to use the `Reshape` and `Transpouse` operators for these actions and then after converting the `Variable` we will have an `Operator` and not a `Variable` with gradients. Also, after changing the form of the `Variable`, we should not return the `Variable`, because it nods a different character in the calculations. What do you think about that?

**Comment by KOHbDS** [comment]
> I don't think `Variable` should have a `gradients` method, because `Variable.gradients` exists only to calculate a gradient of the same other shape using `Transform`. Maybe it is better to use the `Reshape` and `Transpouse` operators for these actions and then after converting the `Variable` we will have an `Operator` and not a `Variable` with gradients. Also, after changing the form of the `Variable`, we should not return the `Variable`, because it nods a different character in the calculations. What do you think about that?

File: tests/apps/test_neural_net.py [hunk 1]
```diff
@@ -14,40 +14,40 @@ class NeuralNetTester(tc.app.Library):
...
+        cxt.optimizer = tc.ml.optimizer.GradientDescent(layer)
         return cxt.optimizer.train(1, inputs, labels)
 
     @tc.post
     def test_cnn(self, cxt, inputs: tc.tensor.Tensor, labels: tc.tensor.Tensor) -> tc.F32:
         layers = [
-            tc.hosted_ml.nn.ConvLayer.create([3, 5, 5], [2, 1, 1], activation=tc.hosted_ml.sigmoid),
-            tc.hosted_ml.nn.ConvLayer.create([2, 3, 3], [2, 1, 1])
+            tc.ml.nn.ConvLayer.create([3, 5, 5], [2, 1, 1], activation=tc.ml.sigmoid),
+            tc.ml.nn.ConvLayer.create([2, 3, 3], [2, 1, 1])
```
Comment (KOHbDS) [inline tests/apps/test_neural_net.py:30]: I think, We have to test it with Adam optimizer.

File: client/tinychain/hosted_ml/optimizer.py [hunk 2]
```diff
@@ -23,24 +24,27 @@ def train(self, inputs):
...
     @post
-    def train(self, inputs: Tensor, labels: Tensor) -> Tensor:
-        variables = trainable(self.ml_model)
+    def train(self, i: UInt, inputs: Tensor, labels: Tensor) -> Tensor:
         outputs = self.ml_model.operator(inputs)
-
         if not isinstance(form_of(outputs), Operator):
             raise ValueError(f"Optimizer can only train a differentiable Operator, not {form_of(outputs)}")
 
         loss = 0.5 * (outputs - labels)**2  # TODO: support an arbitrary cost function
```
Comment (KOHbDS) [inline client/tinychain/hosted_ml/optimizer.py]: if you mean `MSELoss`, I think that look like: `mse =  (outputs - labels)**2`

File: tests/apps/test_neural_net.py (hunk 1 above)
Comment (haydnv) [reply tests/apps/test_neural_net.py:30]: This is a good point, I just found a bug where `Sequential.operator` loses track of the `Variable`s in layer 0 because of the `Tensor.concatenate` in layer 1

File: client/tinychain/math/operator.py [hunk 3]
```diff
@@ -106,7 +106,7 @@ def backward(self, variable):
         return (subject @ self.args) + (self.subject @ arg)
```
[the remark this answers is not in the document]
Comment (haydnv) [reply client/tinychain/math/operator.py:106]: `backward` is useful to calculate the (partial) derivative of an `Operator`, e.g. to calculate the derivative of a cost function automatically, or in a situation where a mathematical model needs to calculate the derivative of its input. I can rename it to `derivative` if you think that's more descriptive. One way I think it might be useful is to define a general solution to calculating the gradients of an `Operator`, which might solve your problem with `Add`. Consider:

```python
def gradients(operator, d_loss):
    if isinstance(operator.subject, Variable):
        return d_loss * operator.backward(operator.subject)
    # ...
```

File: client/tinychain/math/operator.py [hunk 4]
```diff
@@ -78,7 +78,7 @@ def backward(self, variable):
 
     def gradients(self, loss):
```
[the remark this answers is not in the document]
Comment (haydnv) [reply client/tinychain/math/operator.py:79]: An `Adam` optimizer needs to know the namespace of its model's `Variable`s at `__init__` time, which means that a `Variable` which is a transform of another `Variable` needs a method to transform a given `loss` to match the shape of the source `Variable`. You have a point though that `gradients` is maybe not a good name for this because it doesn't actually calculate a gradient, it only inverts a transform. I renamed it to `invert`.

File: tests/apps/test_neural_net.py [hunk 5]
```diff
@@ -14,40 +14,40 @@ class NeuralNetTester(tc.app.Library):
...
+        layer = tc.ml.nn.DNNLayer.create(2, 1, tc.ml.sigmoid)
+        cxt.optimizer = tc.ml.optimizer.Adam(layer)
         return cxt.optimizer.train(1, inputs, labels)
 
     @tc.post
     def test_dnn(self, cxt, inputs: tc.tensor.Tensor, labels: tc.tensor.Tensor) -> tc.F32:
         layers = [
-            tc.hosted_ml.nn.DNNLayer.create(2, 2, tc.hosted_ml.sigmoid),
-            tc.hosted_ml.nn.DNNLayer.create(2, 1, tc.hosted_ml.sigmoid)]
+            tc.ml.nn.DNNLayer.create(2, 2, tc.ml.sigmoid),
```
[the remark this answers is not in the document]
Comment (haydnv) [reply tests/apps/test_neural_net.py]: I added a test for a three-layer `Sequental` deep neural net with irregularly-sized layers

File: tests/apps/test_neural_net.py (hunk 1 above)
Comment (haydnv) [reply tests/apps/test_neural_net.py:30]: I updated the test of a `Sequential` network of `ConvLayer`s to use `Adam` and added a custom `Operator` definition to support backpropagation (I'm not sure if the math is correct but we should be able to tell with further testing)
